### DISCUSSION TITLE
Router is filling in a null object with all the null fields inside the object

### DISCUSSION THREAD

**Discussion by golfingal72** [apollographql/router#1377] [Q&A]
[See my datatypes below] When i return a null on the RetailerOrganization below from a subgraph response, the final response is returning: 
`"retailerOrganization":{"allowedFeatures":null,"retailers":null}`

the APOLLO-SERVER in NODE JS is returning in the final federated response...
`"retailerOrganization":null`

Which is the PREFERRED answer i'd like to see..  

I've tried it 2 ways: not returning the reference to the retailerOrganization in the SUBGRAPH Response AND returning a specific null in the `body: "retailerOrganization": Null`

demonstrated object types:

```graphql
type RetailerOrganization {
  allowedFeatures: [ConnectFeatureContractFeature]
  retailers: [Retailer]
}

type FullUser @key(fields: "id options") @extends {
   id: Long @external
  email: String @external @obfuscateRequestLogging
  options: UserRenderParms @external
  retailerOrganization: RetailerOrganization @requires(fields: "id email")
}
```

**Comment by abernix** [comment]
Is your question in https://github.com/apollographql/router/discussions/1375 the same concern or very closely related to this? (It seems to be?)

**Reply by golfingal72** [reply]
kindof, but not really...  the other issue was an improvement to make outbound payloads smaller...  This is a notable difference between the apollo-service in node.js and this version...  I just updated to your latest version 0.10.0 to see if it goes away and it does not:  

what ends up happening is the json i deliver, i remove null and empty fields and in this case since the RetailerOrganization comes back as an object with all nulls...  show  just an empty object:

 "shopSensePartner": true,
    "type": "Internal",
    "apiPartner": false,
    "retailerOrganization": {},

I have an intermediate gateway between the apollo-router and the client...

The old apollo-server just returned: retailerOrganization: null, NOT retailerOrganization: {element: null, elementa: null}

**Reply by cris-santos** [reply]
@golfingal72 did you find a way to go around this issue? I'm facing the same issue here 😔

**Reply by abernix** [reply]
@cris-santos what version are you using? we have definitely evolved this in the recent months. (please note that 0.10.0 is far from the latest version now even if it was 2 months ago. 😁 )

**Reply by cris-santos** [reply]
@abernix I'm using v1.0.0-alpha.3

<img width="577" alt="Screen Shot 2022-09-14 at 08 05 57" src="https://user-images.githubusercontent.com/7927374/190138204-37b49a59-9fff-4229-b779-f37516663f04.png">
<img width="556" alt="Screen Shot 2022-09-14 at 08 06 48" src="https://user-images.githubusercontent.com/7927374/190138316-a8b7050c-463f-472b-84e1-74964e90f5b8.png">
<img width="865" alt="Screen Shot 2022-09-14 at 08 07 04" src="https://user-images.githubusercontent.com/7927374/190138254-665f852d-eb6b-4aee-befb-68b7d6fbb34e.png">

**Reply by bnjjj** [reply]
@cris-santos could you show me that part of your schema to help me to reproduce and fix it please ?

**Reply by cris-santos** [reply]
@bnjjj thank you for your support!

This is part of the supergraph that would reproduce that error
```graphql
type Query
  @join__type(graph: LOGISTIC_POINT)
  @join__type(graph: ORGANIZATION)
  @join__type(graph: USER) {
  currentUser: User @join__field(graph: USER)
}

type User
  @join__type(graph: ORGANIZATION, key: "id", resolvable: false)
  @join__type(graph: USER, key: "id") {
  id: ID!
  activeOrganization: Organization @join__field(graph: USER)
}

type Organization
  @join__type(graph: LOGISTIC_POINT, key: "id", resolvable: false)
  @join__type(graph: ORGANIZATION, key: "id")
  @join__type(graph: USER, key: "id", resolvable: false) {
  id: ID!
  logisticPoint: LogisticPoint @join__field(graph: ORGANIZATION)
  creatorUser: User @join__field(graph: ORGANIZATION)
}

type LogisticPoint
  @join__type(graph: LOGISTIC_POINT, key: "id")
  @join__type(graph: ORGANIZATION, key: "id", resolvable: false) {
  id: ID!
}
```

The issue is that instead of returning `activeOrganization: null` when we request it with fields that would be resolved from other subgraphs, those fields return `null` and breaks the parent field (`Organization`)

**Reply by bnjjj** [reply]
@cris-santos I think the results you have here are the correct ones. Let me explain, in your schema it seems your `LOGISTIC_POINT` subgraph needs the `ORGANIZATION.id` which in your case is null. Then it's impossible to have data. The rule about bubble up null data is explained here https://spec.graphql.org/October2021/#sel-FAPHRPDTAACVAn-e and then in your second query you're not querying `id` then we don't put `activeOrganization: null`

**Reply by cris-santos** [reply]
@bnjjj, but what about the third query? Is that error supposed to be returned? The third query response seems correct, but the error seems off. In that way, it will always return an error if the organization is null instead of just returning `activeOrganization: null`

**Reply by bnjjj** [reply]
Yes you'll always end up with this error because you're querying `id` for your organization which is a `NonNull` type. As it returns null you have this error because it should not be null

**Reply by cris-santos** [reply]
Hmm, I see. Previously, the apollo gateway with the same schema returned `activeOrganization: null` without any errors. So this new behavior would be considered a breaking change?

**Reply by bnjjj** [reply]
I will follow this to the gateway team, it might be a bug in the gateway. I'll let you know :) feel free to ping me again

**Reply by Geal** [reply]
@cris-santos just to clarify, what is your user subgraph returning for `{ currentUser { activeOrganization { id } }` ?

**Reply by cris-santos** [reply]
@Geal this is the response when I run the query on the user subgraph:
<img width="608" alt="image" src="https://user-images.githubusercontent.com/7927374/190463205-e912cc44-51da-440f-8a08-ca2710996fe7.png">

it's the same response I get when running the same query on the router
<img width="645" alt="image" src="https://user-images.githubusercontent.com/7927374/190464865-befdb340-185a-4b8f-9706-e68c57952289.png">

but if I try to get `cretorUser` for example, it also returns the error
<img width="1018" alt="image" src="https://user-images.githubusercontent.com/7927374/190465488-fc765b9f-feb6-447d-b096-c9ac6ec1ed7d.png">

**Reply by Geal** [reply]
Ok there's probably a bug here, I'll look into it

**Reply by cris-santos** [reply]
@Geal, thank you. If you need more info, just let me know!

**Reply by Geal** [reply]
@cris-santos thanks, I got enough to reproduce it in https://github.com/apollographql/router/pull/1819. I think I know where it comes from, I should have a fix next week
